### PR TITLE
add impersonation_chain support when calling Bigquery Operators in deferrable mode

### PR DESCRIPTION
**Pull request by Lee-W**
* Add missing impersonation_chain argument when calling trigger in Bigquery Operators #36341 
* Simulate BaseToken in gcloud-aio with ImpersonationToken and generate desired token

https://github.com/apache/airflow/issues/34727

---
**^ Add meaningful description above**
Read the **[Pull Request Guidelines](https://github.com/apache/airflow/blob/main/CONTRIBUTING.rst#pull-request-guidelines)** for more information.
In case of fundamental code changes, an Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvement+Proposals)) is needed.
In case of a new dependency, check compliance with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
In case of backwards incompatible changes please leave a note in a newsfragment file, named `{pr_number}.significant.rst` or `{issue_number}.significant.rst`, in [newsfragments](https://github.com/apache/airflow/tree/main/newsfragments).

### CODE REVIEW

**Comment by nathadfield** [comment]
@Lee-W Thanks for taking a look at this but I don't think this will fully solve the problem.  If you follow the trigger call through, it leads to `GoogleBaseHook.provide_gcp_credential_file_as_context()` but there isn't any logic to deal with the impersonation chain.

https://github.com/apache/airflow/blob/main/airflow/providers/google/common/hooks/base_google.py#L499-L532

**Comment by Lee-W** [comment]
I dig it a bit more. It seems the lib we're using does not support Impersonated credential. https://github.com/talkiq/gcloud-aio/issues/421

**Comment by m1racoli** [comment]
So we might first need to look into impersonation chain support in gcloud-aio?

**Comment by nathadfield** [comment]
It's not the most active library to rely upon for a key part of GCP authentication and deferrable task capabilities.

**Comment by Lee-W** [comment]
> @Lee-W Thanks for taking a look at this but I don't think this will fully solve the problem. If you follow the trigger call through, it leads to `GoogleBaseHook.provide_gcp_credential_file_as_context()` but there isn't any logic to deal with the impersonation chain.
> 
> https://github.com/apache/airflow/blob/main/airflow/providers/google/common/hooks/base_google.py#L499-L532

Thanks for reminding me! Yep, I've done some test afterward and find it did not work.

> So we might first need to look into impersonation support in gcloud-aio?

Yep, I think we could take a look at how `gcloud` implement this feature and see what we can do 

> It's not the most active library to rely upon for a key part of GCP authentication and deferrable task capabilities.

I tried to did the investigation but did not find a more active one 😞 aiogoogle does not seem to fit our need either

**Comment by Lee-W** [comment]
after digging a bit deeper, I doubt we can use a similar way to use this impersonation feature https://github.com/googleapis/google-auth-library-python/blob/776d634ac6d989b224f8dbfb11d166cb3025a342/google/auth/_default_async.py#L29

File: airflow/providers/google/cloud/hooks/bigquery.py
Comment (melugoyal) [inline airflow/providers/google/cloud/hooks/bigquery.py]: `impersonation_chain` can have multiple service accounts in the chain. you'll have to handle that by doing something similar to `_get_target_principal_and_delegates`: https://github.com/apache/airflow/blob/e2393ee6dd3a927e753e6375621af07aa0c734dc/airflow/providers/google/common/hooks/base_google.py#L273-L285

Comment (Lee-W) [reply airflow/providers/google/cloud/hooks/bigquery.py]: Thanks for reminding me! just addressed

**Comment by m1racoli** [comment]
FYI We might get better impersonation support in gcloud-aio soon. :)
https://github.com/talkiq/gcloud-aio/pull/665

**Comment by Lee-W** [comment]
> FYI We might get better impersonation support in gcloud-aio soon. :) [talkiq/gcloud-aio#665](https://github.com/talkiq/gcloud-aio/pull/665)

This looks great! Should we use the current PR as a quick fix? Or is it ok to wait for the release of the next gcloud-aio?

**Comment by eladkal** [comment]
Lets wait for a fix in upstream first

**Comment by Lee-W** [comment]
> Lets wait for a fix in upstream first

Sure.

**Review by phanikumv — CHANGES_REQUESTED** [review]
Blocking it till the upstream changes are ready

File: airflow/providers/google/cloud/hooks/bigquery.py
Comment (m1racoli) [inline airflow/providers/google/cloud/hooks/bigquery.py:3279]: Instead of getting default credentials, we should retrieve the credentials for a given connection ID here, no?

Furthermore, I was looking at other Google hooks which work fine asyncronously with impersonation chain and noticed that for example the [DataprocAsyncHook](https://github.com/apache/airflow/blob/9b5d6bfe273cf6af0972e28ff97f99ea325cd991/airflow/providers/google/cloud/hooks/dataproc.py#L1085) (subclass of GoogleBaseHook) just calls `self.get_credentials()`.

I suppose this itself (creating a credentials object) is non-blocking (unless you consider file IO blocking) until we want to actually generate a token for the given credentials, or? So I wonder if we maybe just have to implement `credentials -> token` asynchronously? That's probably what Google's async clients do under the hood, but unfortunately we're here also because there is no async client for [BigQuery](https://cloud.google.com/python/docs/reference/bigquery/latest/google.cloud.bigquery.client.Client) to begin with.

Comment (m1racoli) [reply airflow/providers/google/cloud/hooks/bigquery.py:3279]: I would like to add the `GoogleBaseHook.get_credentials()` probably not only respects impersonation chain set on hook level, but also on connection level. If we can rely on that, then we would need to cover those cases individually.

File: airflow/providers/google/cloud/hooks/bigquery.py
Comment (m1racoli) [inline airflow/providers/google/cloud/hooks/bigquery.py:3288]: This need's to be done asynchronously. Otherwise it will block the entire triggerer process.

**Review by m1racoli — CHANGES_REQUESTED** [review]

**Comment by m1racoli** [comment]
Based on my review comments I've taken another attempt on this in #36849. That one should make it easier to use credentials from hooks in gcloud-aio clients without much extra work.

**Comment by Lee-W** [comment]
Hi @m1racoli, thanks for your feedback! I just took a look at your new PRs. Both of them look good. I think we might no longer need this PR, but I'll mark it as draft for now before we merge yours

**Comment by Lee-W** [comment]
as https://github.com/apache/airflow/pull/36849 has been merged, close this one

**Pull request closed without merging**
